File: 0155. Min Stack.cpp
//Time Complexity: O(1)
//Space Complexity: O(n)
//Approach: Prefix Min, Stacks
class MinStack {
public:
    int t;
    vector<int> stack;
    vector<int> mins;
    MinStack() {
        t = -1;
    }
    
    void push(int val) {
        stack.push_back(val);
        if(t >= 0) mins.push_back(min(val, mins[t]));
        else mins.push_back(val);
        t++;
    }
    
    void pop() {
        stack.erase(stack.begin() + t);
        mins.erase(mins.begin() + t);
        t--;
    }
    
    int top() {
        return stack[t];
    }
    
    int getMin() {
        return mins[t];
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155. Min Stack.py
# Time Complexity: O(1) for all ops
# Space Complexity: O(n)
# Approch: Stack
class MinStack:
    def __init__(self):
        self.stack = []

    # Every element keeps tracks of the min element up to that point
    def push(self, val: int) -> None:
        if len(self.stack) == 0:
            self.stack.append([val, val])
        else:
            self.stack.append([val, min(val, self.stack[-1][1])])

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        return self.stack[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
